plugins/outputs/azuremonitor/azuremonitor.go

```diff
@@ -0,0 +1,445 @@
+	}
+
+	if !a.useMsi {
+		// If using direct AD authentication create the AD access client
+		oauthConfig, err := adal.NewOAuthConfig(azure.PublicCloud.ActiveDirectoryEndpoint, a.AzureTenantID)
+		if err != nil {
+			return fmt.Errorf("E! Could not initialize AD client: %s", err)
+		}
+		a.oauthConfig = oauthConfig
+	}
+
+	// Pull region and resource identifier
+	err := a.GetInstanceMetadata()
+	if err != nil && a.ResourceID == "" && a.Region == "" {
+		return fmt.Errorf("E! No resource id specified, and Azure Instance metadata service not available.  If not running on an Azure VM, provide a value for resource_id")
+	}
+
+	err = a.validateCredentials()
+	if err != nil {
+		return fmt.Errorf("E! Unable to fetch authentication credentials: %v", err)
+	}
+
+	a.Reset()
+
+	return nil
+}
+
+func (a *AzureMonitor) validateCredentials() error {
+	if a.useMsi {
+		// Check expiry on the token
+		if a.msiToken == nil || a.msiToken.expiresInDuration() < time.Minute {
+			msiToken, err := a.getMsiToken(a.AzureClientID)
+			if err != nil {
+				return err
+			}
+			a.msiToken = msiToken
+		}
+		return nil
+	}
+
+	adToken, err := adal.NewServicePrincipalToken(
+		*(a.oauthConfig), a.AzureClientID, a.AzureClientSecret,
+		azure.PublicCloud.ActiveDirectoryEndpoint)
+	if err != nil {
+		return fmt.Errorf("Could not acquire ADAL token: %s", err)
+	}
+	a.adalToken = adToken
+	return nil
+}
+
+// Close shuts down an any active connections
+func (a *AzureMonitor) Close() error {
+	a.client = nil
+	return nil
+}
+
+type azureMonitorMetric struct {
+	Time time.Time         `json:"time"`
+	Data *azureMonitorData `json:"data"`
+}
+
+type azureMonitorData struct {
+	BaseData *azureMonitorBaseData `json:"baseData"`
+}
+
+type azureMonitorBaseData struct {
+	Metric         string                `json:"metric"`
+	Namespace      string                `json:"namespace"`
+	DimensionNames []string              `json:"dimNames"`
+	Series         []*azureMonitorSeries `json:"series"`
+}
+
+type azureMonitorSeries struct {
+	DimensionValues []string `json:"dimValues"`
+	Min             float64  `json:"min"`
+	Max             float64  `json:"max"`
+	Sum             float64  `json:"sum"`
+	Count           int64    `json:"count"`
+}
+
+// Write writes metrics to the remote endpoint
+func (a *AzureMonitor) Write(metrics []telegraf.Metric) error {
+	azmetrics := make(map[uint64]*azureMonitorMetric, len(metrics))
+	for _, m := range metrics {
+		id := hashIDWithTagKeysOnly(m)
+		if azm, ok := azmetrics[id]; !ok {
+			azmetrics[id] = translate(m)
+		} else {
+			azmetrics[id].Data.BaseData.Series = append(
+				azm.Data.BaseData.Series,
+				translate(m).Data.BaseData.Series...,
+			)
+		}
+	}
+
+	var body []byte
+	for _, m := range azmetrics {
+		// Azure Monitor accepts new batches of points in new-line delimited
+		// JSON, following RFC 4288 (see https://github.com/ndjson/ndjson-spec).
+		jsonBytes, err := json.Marshal(&m)
```

Comment:
How large can azmetrics be?  I believe the Azure Monitor metric API has a max request body size of 4MB.  If we exceed this limit, we should issue multiple POST requests